### PR TITLE
US104726 - Retain number of shown activities on sort

### PR DESCRIPTION
**Pull request by mtjandra**
**Note:** Demo only works after https://github.com/BrightspaceHypermediaComponents/activities/pull/124

retainNumActivities

* Retain number of activities shown when sorting

### CODE REVIEW

File: components/d2l-quick-eval/d2l-quick-eval-activities-list.js [hunk 1]
```diff
@@ -418,7 +426,8 @@ class D2LQuickEvalActivitiesList extends mixinBehaviors([D2L.PolymerBehaviors.Si
 				return action;
 			}).bind(this))
 			.then((collectionAction => {
-				const collection = this._performSirenActionWithQueryParams(collectionAction);
+				const customParams = this._numberOfCurrentlyShownActivities > 0 ? {pageSize: this._numberOfCurrentlyShownActivities} : {};
+				const collection = this._performSirenActionWithQueryParams(collectionAction, customParams);
```
Comment (mtjandra) [inline components/d2l-quick-eval/d2l-quick-eval-activities-list.js]: HREF on web-component never gets updated. So this `pageSize` param should not be propagated to filters or main WC. I tested on LMS as well.

File: test/d2l-quick-eval/d2l-quick-eval-activities-list.js [hunk 2]
```diff
@@ -65,6 +65,15 @@ import '@polymer/iron-test-helpers/mock-interactions.js';
 		done();
 	}
 
+	function arrayContainsWhere(arr, f) {
```
Comment (AlexBedley) [inline test/d2l-quick-eval/d2l-quick-eval-activities-list.js]: I think even ie supports [`some`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some) (although ironically it does not support `includes`)

File: test/d2l-quick-eval/d2l-quick-eval-activities-list.js [hunk 3]
```diff
@@ -410,6 +419,27 @@ import '@polymer/iron-test-helpers/mock-interactions.js';
...
+
+			const action = {
+				href : 'http://127.0.0.1/',
+				name: 'apply',
+				type: 'application/x-www-form-urlencoded',
+				method: 'GET'
+			};
+
+			const customParams = { customParam1: 'custom', customParam2: 'custom2' };
+			sinon.stub(list, 'performSirenAction', function(passedAction) {
```
Comment (AlexBedley) [inline test/d2l-quick-eval/d2l-quick-eval-activities-list.js]: nit: I don't mind the current implementation but `sinon` has some cool built-ins you could use. You seem to be familiar with the [stub API](https://sinonjs.org/releases/latest/stubs/) but there's also the [spy API](https://sinonjs.org/releases/latest/spies/)

For example:

```js
const expectedAction = ...; // whatever
const spy = sinon.spy(list, 'performSirenAction');

list._performSirenActionWithQueryParams(action, customParams);

assert(spy.withArgs(expectedAction).calledOnce);
```

File: components/d2l-quick-eval/d2l-quick-eval-activities-list.js [hunk 4]
```diff
@@ -418,7 +426,8 @@ class D2LQuickEvalActivitiesList extends mixinBehaviors([D2L.PolymerBehaviors.Si
 				return action;
 			}).bind(this))
 			.then((collectionAction => {
-				const collection = this._performSirenActionWithQueryParams(collectionAction);
+				const customParams = this._numberOfCurrentlyShownActivities > 0 ? {pageSize: this._numberOfCurrentlyShownActivities} : {};
```
Comment (AlexBedley) [inline components/d2l-quick-eval/d2l-quick-eval-activities-list.js]: What case does the `> 0` case handle? If we were sorting from an error state?

nit: Also `{}` evaluates to true so we are entering that `if (customParams)` block in the function below

File: components/d2l-quick-eval/d2l-quick-eval-activities-list.js [hunk 5]
```diff
@@ -261,6 +261,10 @@ class D2LQuickEvalActivitiesList extends mixinBehaviors([D2L.PolymerBehaviors.Si
 				type: Array,
 				value: [ ]
 			},
+			_numberOfCurrentlyShownActivities: {
+				type: Number,
+				computed: '_computeNumberOfCurrentlyShownActivities(_data)'
```
Comment (AlexBedley) [inline components/d2l-quick-eval/d2l-quick-eval-activities-list.js]: Curious as to why you went this way vs just using `data.length` everywhere (I don't know anything about polymer `computed` properties)

File: test/d2l-quick-eval/d2l-quick-eval-activities-list.js (hunk 2 above)
Comment (mtjandra) [reply test/d2l-quick-eval/d2l-quick-eval-activities-list.js]: Thank you!!! I will this out

File: test/d2l-quick-eval/d2l-quick-eval-activities-list.js (hunk 3 above)
Comment (mtjandra) [reply test/d2l-quick-eval/d2l-quick-eval-activities-list.js]: Thanks for the reference! Although I'm not sure what this buys us. Though I do notice now that I probably want to add a `done()` within the stub.

File: components/d2l-quick-eval/d2l-quick-eval-activities-list.js (hunk 4 above)
Comment (mtjandra) [reply components/d2l-quick-eval/d2l-quick-eval-activities-list.js]: > What case does the > 0 case handle? If we were sorting from an error state?

Yes I was thinking if say we received an error on load, so there is no information in the data. Then someone clicks one of the header, and the data loads correctly. I wouldn't want to maintain showing nothing, but instead whatever the default is.

> nit: Also {} evaluates to true so we are entering that if (customParams) block in the function below

 This is why I dislike JS. I will change it to undefined thanks

File: components/d2l-quick-eval/d2l-quick-eval-activities-list.js (hunk 5 above)
Comment (mtjandra) [reply components/d2l-quick-eval/d2l-quick-eval-activities-list.js]: I just wanted to name `data.length` as `_numberOfCurrentlyShownActivities`. I doubt the meaning of data.length will ever change, but just incase it does. It seemed to make sense in my head that our component knows how many activities are shown, so a property seemed good. Might just be personal preference though

File: test/d2l-quick-eval/d2l-quick-eval-activities-list.js [hunk 6]
```diff
@@ -455,6 +476,44 @@ import '@polymer/iron-test-helpers/mock-interactions.js';
...
+				fields : defaultParams
+			};
+
+			const customParams = { customParam1: 'custom', customParam2: 'custom2' };
+			sinon.stub(list, 'performSirenAction', function(passedAction) {
+				const fields = passedAction.fields;
+				assert.equal(4, fields.length);
+
+				assert.isTrue(fields.some(function(elm) { return elm.name === 'testname' && elm.value === 'testvalue'; }));
+				assert.isTrue(fields.some(function(elm) { return elm.name === 'anothertestname' && elm.value === 'anothertestvalue'; }));
```
Comment (mtjandra) [inline test/d2l-quick-eval/d2l-quick-eval-activities-list.js]: `[1, 2, 3].some((e) => e === 2)` throws in IE11 console. I'm not even sure how the test suites run.... but that is why I didn't use anonymous functions

File: test/d2l-quick-eval/d2l-quick-eval-activities-list.js (hunk 3 above)
Comment (AlexBedley) [reply test/d2l-quick-eval/d2l-quick-eval-activities-list.js]: Hmm yeah it's not a revolutionary refactor - I just find using the built-ins more readable. One equality assertion is likely easier to reason with than a series of custom assertions. But like I said, I'm ok with the current implementation.

File: test/d2l-quick-eval/d2l-quick-eval-activities-list.js (hunk 6 above)
Comment (AlexBedley) [reply test/d2l-quick-eval/d2l-quick-eval-activities-list.js]: Super weird, I wonder if some transpiling is going on at a certain level. the magic of `polymer test`

**Pull request merged**